datalad_next/runners/run.py

```diff
@@ -0,0 +1,166 @@
+                self.kill_counter += 1
+
+            # If we set the timeout argument due to a not None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+        cmd: list,
+        protocol_class: type[Protocol],
+        *,
+        cwd: Path | None = None,
+        input: int | IO | bytes | Queue[bytes | None] | None = None,
+        timeout: float | None = None,
+        terminate_time: int | None = None,
+        kill_time: int | None = None,
+) -> Any | Generator:
+
+    introduces_timeout = False
+    if timeout is None:
+        introduces_timeout = True
+        timeout = 1.0
+
+    armed = (
+        False
+        if issubclass(protocol_class, GeneratorMixIn)
+        else True
+    )
+
+    # Create the wrapping class. This is done mainly to ensure that the
+    # termination-related functionality is present in the protocol class that
+    # is used, independent of the actual protocol class that the user passes as
+    # argument.
+    # A side effect of this approach is, that the number of protocol class
+    # definitions is reduced, because the user does not need to define
+    # terminate-capable protocols for every protocol he wants to use.
+    kill_protocol_class = create_kill_wrapper(protocol_class)
+
+    runner = ThreadedRunner(
+        cmd=cmd,
+        protocol_class=kill_protocol_class,
+        stdin=DEVNULL if input is None else input,
+        protocol_kwargs=dict(
+            introduced_timeout=introduces_timeout,
+            terminate_time=terminate_time,
+            kill_time=kill_time,
+            armed=armed,
+            done_future=None,
+            encoding=None
+        ),
+        timeout=timeout,
+        exception_on_error=False,
+        cwd=cwd,
+    )
+    result = runner.run()
+    # We distinguish between a non-generator run, i,e. a blocking run and
+    # a generator run.
+    if not issubclass(protocol_class, GeneratorMixIn):
+        yield result
+    else:
+        try:
+            yield result
+        finally:
+            # Arm the protocol, that will enable terminate signaling or kill
+            # signaling, if terminate_time or kill_time are not None.
+            result.runner.protocol.arm()
+            # Exhaust the generator. Because we have set a timeout, this will
+            # lead to invocation of the timeout method of the wrapper, which
+            # will take care of termination or killing. And it will fetch
+            # the result code of the terminated process.
+            # NOTE: if terminate_time and kill_time are both None, this might
+            # loop forever.
```

Comment:
Something to protect against? Surely documentation is needed.

Reply:
I am not sure how to protect against that. A user might intentionally not specify a terminate or kill time in order to let an important process, e.g. database-vacuum, do its job. On the other hand, a user might just not think about this and unintentionally let the process run forever. We would not be able to distinguish those two situations

Documentation and examples will go into the run-doc string and the "Runner Developers Handbook".

Reply:
Dcoumentation is also added to the doc-string of the context manager